`spack find -c`: search all concretized-but-not-installed specs

scheibelp opened the pull request:
Originally if you had `x -> y -> z`, and an env with `x` in its speclist that is concretized but not installed, then `spack find -c y` would not show anything. This was intended: `spack find` has up-until-now only ever listed out installed specs (and `-c` was for adding a preamble section about roots).

This changes `spack find` so:

* `-c` makes it search through all concretized specs in the env (in a sense it is anticipated that a concretized environment would serve as a "speculative" DB and users may want to query it like they query the DB outside of envs)
* Adds a `-i/--install-status` option, equivalent to `-I` from `spack spec`
* Shows install status for either `-c` or `-i`
* As a side effect to prior point, `spack find -i` can now distinguish different installation states (upstream/external)

The overall effect looks like:

```
$ spack find -c 
==> In environment findtest
==> 1 root specs
 -  raja

==> Concretized roots
-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
raja@2024.02.2

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
[+]  berkeley-db@18.1.40   -   camp@2024.02.1  [+]  diffutils@3.10  [+]  gnuconfig@2022-09-17   -   nghttp2@1.62.0   -   raja@2024.02.2
 -   blt@0.6.2             -   cmake@3.29.4     -   gdbm@1.23       [+]  libiconv@1.17          -   perl@5.38.2      -   readline@8.2
[+]  bzip2@1.0.8           -   curl@8.7.1      [+]  gmake@4.4.1      -   ncurses@6.5            -   pkgconf@2.2.0    -   zlib-ng@2.1.6
==> 12 concretized-but-not-installed packages
==> 6 installed packages
```

Notes:

- `spack find -d` does not consolidate duplicate dependencies like https://github.com/spack/spack/pull/44489. I think this would be straightforward but I'm not sure if that's what users would want with `-d`

scheibelp committed:
partial support for searching an active env for concretized-but-not-installed specs

scheibelp committed:
Merge branch 'develop' into features/find-env-concretized

scheibelp committed:
properly report concretized-but-not-installed specs when using -c

scheibelp committed:
install status for find (e.g. whether an installed spec is external or not

scheibelp committed:
add type to Spec.install_status, shorten colorize logic in display_specs

scheibelp committed:
now that we have install status, don't break down installed/not-installed into separate sections

scheibelp committed:
style edit

scheibelp committed:
update docstring w/ new kwarg

scheibelp committed:
install-statu is true by default for 'spack spec', but should be false by default for 'spack find'

scheibelp committed:
update shell completion

scheibelp commented:
> No thought went into the `spack -e <env> find -c`

Not sure what you mean: running `spack -e ... find -c` when an env is otherwise inactive is just like running `spack find -c` while an env is active; that follows the general behavior of `spack -e ...`

> In my opinion we better discuss what the semantics should be before incrementally making the command increasingly different for environments.
> ...
> I think generally it's OK to make spack -e <env> find search through the concrete environment, disregarding install status.

My choice to use `-c` for searching uninstalled specs was an attempt at resolving the dissonance between these two statements: you don't get uninstalled specs unless you specifically enable an option that says you want to see them.

> useful, but nobody will remember to add a -c flag. 

but (as a continuation of prior point), I see it would be nice to fold in, because it would not necessarily be obvious when users forgot `-c` ("did I forget to concretize my environment?"), except...

> Calling specs "concretized-but-not-installed" looks very ugly, I wouldn't give it a special status.

this message would help (and I could add "not shown", when `-c` is missing).

Overall I slightly prefer `spack find -c` (because it generally preserves the behavior of `spack find` searching only-installed specs). I think `spack find` automatically doing this in envs could work though (in particular as long as the install status is always shown).

> Also it no longer prints installed specs only.

I felt that making `-c` activate the install status would help clarify that: the format is much like `spack spec`, so I don't think it would be confusing to users.

> and currently there is simply no way to run `spack -e <env>` find to figure out what's in your db in a custom location:

To make sure I understand the problem: `-e` uses the env and constrains all results to only be specs from the env, and you want to be able find specs that may or may not be associated with the env that are installed inside of the `install_root` which is provided by the env.

I would prefer an option like `spack find --do-not-constrain-by-env`: very specific, but IMO that is merited since the context is specific. Either way, if this can exist as an option on `spack find` vs. `spack`, then I think it can be handled separately (I don't mind discussing it here though).

tgamblin commented:
> No thought went into the spack -e <env> find -c option, it was just there cause there was no way to get spack concretize output after concretization.

I am not sure how you can know that "no thought" went into it, or why it matters at this point, but I agree it could be improved. I think the latter is more important.

> spack find used to be a query on installed packages in the database.

> We've discussed earlier to merge spack buildcache list into spack find, which basically runs a query against build cache database merged with the usual upstreams + user db -- that violates spack find's original use of listing installed packages as well.

I don't really agree with this assumption about what `space find` is for.  `spack find` and `spack buildcache list` (which IMO should either be called `space buildcache find` or merged into `spack find`) are for *querying and filtering sets of concrete specs*.  Looking at installed specs is a subset of that.  But filtering concrete specs is basically what we're doing for all the use cases you mention:

* listing/querying/filtering installed packages
* listing/querying/filtering concrete packages to be installed
* listing/querying/filtering buildcache specs that could be installed

We can query concrete specs in an upstream spack instance with `--install-tree TREE`... I don't see a huge difference between that and adding a `--buildcache` option.  The options you'd want to use to query in both cases are the same.

`spack spec` is for printing the results of concretization and telling you which parts are installed already. The two are related somewhat but I don't think it's awful that we have a tree view for `spack spec` and a query view for other cases.

> I think generally it's OK to make spack -e <env> find search through the concrete environment, disregarding install status.

Ok I think we all agree on this then.

> I don't think it makes sense to hide that behind a flag -c: spack find would then just mean "find something in my environment".

I'm not opposed to showing concrete stuff by default... but I think it only really makes sense if you show install status, which makes the default behavior of `spack find` even more different from how it behaves outside an env.  What if we had output like this:

```
==> 6 installed packages
==> 12 packages to be installed (show with `spack find --show-concretized`)
```

That would at least make it easy for a user to discover.

> Lastly, I would like it when I could opt out of this spack find env aware mode. Say spack find -E, like spack -E. The reason is that I use config:install_tree:root:... in environments regularly, and currently there is simply no way to run spack -e <env> find to figure out what's in your db in a custom location:

I don't think I understand your problem here.  Does `spack -E find` not do what you want?

tgamblin commented on lib/spack/spack/cmd/find.py:
```diff
@@ -293,23 +297,27 @@ def root_decorator(spec, string):
         print()
 
-    if args.show_concretized:
+    if args.show_concretized and not args.constraint:
         tty.msg("Concretized roots")
         cmd.display_specs(env.specs_by_hash.values(), args, decorator=decorator)
         print()
```
I think we should just remove the concretized roots section entirely, and just make `-c` show all concrete specs in the main output.  I agree with @haampie that it makes the output weird and it's kind of a one-off section.  The roots are shown in the main output and are already highlighted, so we basically show each concretized root twice.

If you can't see the highlighting (it's subtle in some terminals) we could consider underlining them instead of bolding them if folks think that is better.

tgamblin commented on lib/spack/spack/cmd/find.py:
```diff
@@ -293,23 +297,27 @@ def root_decorator(spec, string):
+
+    if args.show_concretized:
+        if not env:
+            tty.die("-c/specs given but no active environment to search")
```
make this consistent with the other error message below.
```suggestion
             tty.die("-c / --show-concretized requires an active environment")
```

tgamblin commented on lib/spack/spack/cmd/find.py:
```diff
@@ -342,10 +355,20 @@ def find(parser, args):
         if not args.only_roots:
-            cmd.display_specs(results, args, decorator=decorator, all_headers=True)
             count_suffix = ""
+            kwargs = {"decorator": decorator, "all_headers": True, "status_fn": status_fn}
+            cmd.display_specs(results, args, **kwargs)
```
I don't see why we need to change the invocation style here -- IMO this is clearer:

```suggestion
            kwargs = {"decorator": decorator, "all_headers": True, "status_fn": status_fn}
            cmd.display_specs(
                results,
                args,
                decorator=decorator,
                all_headers=True,
                status_fn=status_fn,
            )
```

tgamblin commented on lib/spack/spack/cmd/find.py:
```diff
@@ -342,10 +355,20 @@ def find(parser, args):
 
         # print number of installed packages last (as the list may be long)
         if sys.stdout.isatty() and args.groups:
+            if args.show_concretized:
+                spack.cmd.print_how_many_pkgs(
+                    list(x for x in results if not x.installed),
+                    "concretized-but-not-installed",
+                    suffix=count_suffix,
+                )
```
see my other comment -- I think this needs to be reworded (like @haampie said) and that it should hint that you can use `--show-concretized` to see them.

scheibelp replied:
Harmen didn't want this part at all, regardless of wording. It sounds like you want this, but worded differently. I will change the wording.

tgamblin replied:
Still interested in whether @haampie likes my suggestion but yes this is what I was suggesting.

scheibelp committed:
some review changes

scheibelp committed:
auto-style fix

scheibelp committed:
partial reorg

scheibelp committed:
more partial reorg

scheibelp committed:
I think the suffixes are sorted now

scheibelp commented:
Requests addressed. I also updated the final count info to mention whether you are in an env or not. Examples:

```
bash-3.2$ spack find -r
==> In environment findtest
==> 1 root specs
 -  raja

==> 6 installed packages in the environment (not shown)
==> 12 concretized (not installed) packages in the environment (not shown)



bash-3.2$ spack find -c
==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
[+]  berkeley-db@18.1.40   -   camp@2024.02.1  [+]  diffutils@3.10  [+]  gnuconfig@2022-09-17   -   nghttp2@1.62.0   -   raja@2024.02.2
 -   blt@0.6.2             -   cmake@3.29.4     -   gdbm@1.23       [+]  libiconv@1.17          -   perl@5.38.2      -   readline@8.2
[+]  bzip2@1.0.8           -   curl@8.7.1      [+]  gmake@4.4.1      -   ncurses@6.5            -   pkgconf@2.2.0    -   zlib-ng@2.1.6
==> 6 installed packages in the environment
==> 12 concretized (not installed) packages in the environment



bash-3.2$ spack find
==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
berkeley-db@18.1.40  bzip2@1.0.8     cmake@3.29.4  diffutils@3.10  gmake@4.4.1           libiconv@1.17  nghttp2@1.62.0  pkgconf@2.2.0   readline@8.2
blt@0.6.2            camp@2024.02.1  curl@8.7.1    gdbm@1.23       gnuconfig@2022-09-17  ncurses@6.5    perl@5.38.2     raja@2024.02.2  zlib-ng@2.1.6
==> 6 installed packages in the environment
==> 12 concretized (not installed) packages in the environment (not shown, display with -c)


bash-3.2$ spack -E find
...
==> 82 installed packages
```

tgamblin commented:
>  I also updated the final count info to mention whether you are in an env or not.

Since we already have this:
```
==> In environment findtest
```

I think adding "in the environment" to both counts is redundant and makes this harder to read.

How about:
```
==> 6 installed packages
==> 12 concretized packages to be installed (show with `spack find -c`)
```

scheibelp committed:
fiddling with wording

scheibelp committed:
no more -c for ==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
[+]  berkeley-db@18.1.40
 -   blt@0.6.2
[+]  bzip2@1.0.8
 -   camp@2024.02.1
 -   cmake@3.29.4
 -   curl@8.7.1
[+]  diffutils@3.10
 -   gdbm@1.23
[+]  gmake@4.4.1
[+]  gnuconfig@2022-09-17
[+]  libiconv@1.17
 -   ncurses@6.5
 -   nghttp2@1.62.0
 -   perl@5.38.2
 -   pkgconf@2.2.0
 -   raja@2024.02.2
 -   readline@8.2
 -   zlib-ng@2.1.6; fix test that assumed ==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
[+]  berkeley-db@18.1.40
 -   blt@0.6.2
[+]  bzip2@1.0.8
 -   camp@2024.02.1
 -   cmake@3.29.4
 -   curl@8.7.1
[+]  diffutils@3.10
 -   gdbm@1.23
[+]  gmake@4.4.1
[+]  gnuconfig@2022-09-17
[+]  libiconv@1.17
 -   ncurses@6.5
 -   nghttp2@1.62.0
 -   perl@5.38.2
 -   pkgconf@2.2.0
 -   raja@2024.02.2
 -   readline@8.2
 -   zlib-ng@2.1.6 only dumped installed specs

scheibelp committed:
update completion

scheibelp commented:
> I think adding "in the environment" to both counts is redundant and makes this harder to read.
> How about:

Sounds good

Except I forgot we wanted to get rid of `-c`, so new output:

```
bash-3.2$ spack find -r
==> In environment findtest
==> 1 root specs
 -  raja

==> 6 installed packages (not shown)
==> 12 concretized packages to be installed (not shown)



bash-3.2$ spack find
==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
[+]  berkeley-db@18.1.40   -   camp@2024.02.1  [+]  diffutils@3.10  [+]  gnuconfig@2022-09-17   -   nghttp2@1.62.0   -   raja@2024.02.2
 -   blt@0.6.2             -   cmake@3.29.4     -   gdbm@1.23       [+]  libiconv@1.17          -   perl@5.38.2      -   readline@8.2
[+]  bzip2@1.0.8           -   curl@8.7.1      [+]  gmake@4.4.1      -   ncurses@6.5            -   pkgconf@2.2.0    -   zlib-ng@2.1.6
==> 6 installed packages
==> 12 concretized packages to be installed



bash-3.2$ spack -E find
...
==> 82 installed packages
```

tgamblin commented:
> Except I forgot we wanted to get rid of `-c`, so new output:

I think we still want `-c` to be required to show not-installed specs for reasons outlined above:

> I'm not opposed to showing concrete stuff by default... but I think it only really makes sense if you show install status, which makes the default behavior of spack find even more different from how it behaves outside an env.

I think it makes more sense to *not* do `-c` by default so that only installed specs are shown by default and no special status is needed (otherwise the main output section is even more different in envs).

@haampie what do you thinK?

scheibelp commented:
@spackbot help

scheibelp commented:
Close/reopen to retry docs build

scheibelp commented:
(current unit test failures are related to removal of `-c`, so I'll wait to see how that conversation resolves before fixing)

haampie commented:
> I don't really agree with this assumption about what space find is for.

There is no assumption about this, the command advertises

```
$ spack find --help
usage: spack find ...
list and search installed packages  # <---
```

since well over a decade https://github.com/spack/spack/commit/e645bb065ac790abeda9919e82e0c5808ed6ba3d, so if you want spack find to mean

> querying and filtering sets of concrete specs

it *is* new behavior as of this PR. Before this PR, even with a flag, `spack find -c pkg` would not find a concrete *non-installed* `pkg`.

I'm not opposed to change that, but I would try harder to get it right.

> concrete packages to be installed

I would not call it "concrete packages to be installed", going forward it's unlikely we'll install build deps. Just call it "uninstalled packages". Or "10 of 20 packages installed"

> Does spack -E find not do what you want?

No, I need `spack.yaml` as a config scope for `config:install_tree:root:...`. Example use case is `spack -e . find -El x` to get a hash followed by `spack -e . add y ^x/hash` and `spack -e . concretize`. Might also apply to build caches. That's why I don't like that `spack find` has completely different behavior when an env is activated.

If you decide `spack buildcache find` should remain a separate command, then I would say let's move all the environment specific stuff from `spack find` into `spack env find` instead.

If you want to merge `spack buildcache list` into `spack find`, then certainly we need a way to opt out of the environment restriction to search, cause I might want to use the command to see what is in a buildcache, obviously not filtered by what is in my env, nor what is installed in the database.

At this point I wouldn't even prefer separate commands `spack buildcache find`, `spack env find` and `spack find`.

tgamblin commented:
Per the discussion today:

1. Envs are two things:
    a. Configuration
    b. A limited spec context, defined by a set of roots
2. `find` is a number of things:
    a. a way to query installed packages
    b. a way to query packages in a buildcache/upstream
    c. a way to query concretized but not yet installed specs in an env.
3. (2) gets in the way of (1) inside an environment because you have to take the configuration *and* the spec constraints together. e.g, if you have:
    ```yaml
    spack:
        upstreams: ...
        mirrors: ...
        specs: []
    ```
    You can't use, e.g., `spack find --install-tree NAME` to poke around in the configured upstreams because results are constrained by the roots (and there are none so no results are shown).  Even in a concretized env, you can only see those specs from the upstream picked up by the concretizer.  `spack -E find ...` is not ideal b/c it throws the configuration from the env out along with the spec constraint. We want a way to *just* use the config from the environment.

So we discussed a couple things:
1. `spack -C <dir>` has been around since before environments. It currently takes a raw configuration directory. It should also be able to take an environment name or dir as an argument, and it should *just* use the config from the env (vs. `spack -e <env or dir>` which uses both config and spec constraints)
2. There should probably also be a way to omit the spec constraints when the env is activated. e.g., `spack -S/--no-env-specs`, or perhaps an arg on `find`, that can be used inside an active env.  This would be a power user feature so I'm not super concerned about what it's called but we should try to make it consistent.

@haampie is going to do a separate PR for the modifications to the `-C` option.

We also agreed that:
- `-c` should not be default and should remain as a flag.
- by default `spack find` in an env should show the output we're used to
- `-c`  should imply install status and should show all concrete specs inline with installed specs.

@scheibelp was going to implement this stuff then this PR is good to go.

scheibelp committed:
Revert "update completion"

This reverts commit 09930ad3feeb6630aff872c26e439924152ebc61.

scheibelp committed:
Revert "no more -c for ==> In environment findtest"

This reverts commit 76f3b883f8a300e26c2a605dc4d6074da79178bb.

scheibelp committed:
dont show not-installed specs for envs unless user specifies -c

scheibelp commented:
Updates address requests from https://github.com/spack/spack/pull/44713#issuecomment-2206881517:

```
$ spack find -r
==> In environment findtest
==> 1 root specs
 -  raja

==> 6 installed packages (not shown)
==> 12 concretized packages to be installed (not shown)



$ spack find
==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
berkeley-db@18.1.40  bzip2@1.0.8  diffutils@3.10  gmake@4.4.1  gnuconfig@2022-09-17  libiconv@1.17
==> 6 installed packages
==> 12 concretized packages to be installed (show with `spack find -c`)



$ spack find -c
==> In environment findtest
==> 1 root specs
 -  raja

-- darwin-ventura-m1 / apple-clang@14.0.3 -----------------------
[+]  berkeley-db@18.1.40  [+]  bzip2@1.0.8      -   cmake@3.29.4  [+]  diffutils@3.10  [+]  gmake@4.4.1           [+]  libiconv@1.17   -   nghttp2@1.62.0   -   pkgconf@2.2.0    -   readline@8.2
 -   blt@0.6.2             -   camp@2024.02.1   -   curl@8.7.1     -   gdbm@1.23       [+]  gnuconfig@2022-09-17   -   ncurses@6.5     -   perl@5.38.2      -   raja@2024.02.2   -   zlib-ng@2.1.6
==> 6 installed packages
==> 12 concretized packages to be installed



$ spack -E find
...
==> 82 installed packages
```
